[StaticRuntime] Add a 'speed of light' benchmark.

ZolotukhinM opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#46308 [StaticRuntime] Add a 'speed of light' benchmark.**

This PR adds a hand optimized version of DeepAndWide model with the goal
of estimating overheads of static runtime. While static runtime is
currently much faster than the existing JIT interpreter, it would be
useful to understand how close we are to an absolutely 0-overhead
system. Currently, this "ideal" implementation is 2x faster than the
static runtime on batchsize=1.

Full benchmark results:
```
Running build/bin/static_runtime_bench
Run on (24 X 2394.71 MHz CPU s)
CPU Caches:
  L1 Data 32K (x24)
  L1 Instruction 32K (x24)
  L2 Unified 4096K (x24)
  L3 Unified 16384K (x24)
------------------------------------------------------------------------------
Benchmark                                       Time           CPU Iterations
------------------------------------------------------------------------------
BM_deep_wide_base/1                         59518 ns      59500 ns      10909
BM_deep_wide_base/8                         74635 ns      74632 ns       9317
BM_deep_wide_base/20                        82186 ns      82147 ns       9119
BM_deep_wide_fast/1                         13851 ns      13851 ns      49825 << new
BM_deep_wide_fast/8                         22497 ns      22497 ns      32089 << new
BM_deep_wide_fast/20                        23868 ns      23841 ns      31184 << new
BM_deep_wide_jit_graph_executor/1           62786 ns      62786 ns      10835
BM_deep_wide_jit_graph_executor/8           76730 ns      76718 ns       7529
BM_deep_wide_jit_graph_executor/20          78886 ns      78883 ns       8769
BM_deep_wide_jit_profiling_executor/1       69504 ns      69490 ns      10309
BM_deep_wide_jit_profiling_executor/8       75718 ns      75715 ns       9199
BM_deep_wide_jit_profiling_executor/20      75364 ns      75364 ns       9010
BM_deep_wide_static/1                       40324 ns      40318 ns      17232
BM_deep_wide_static/8                       50327 ns      50319 ns      13335
BM_deep_wide_static/20                      53075 ns      53071 ns      12855
BM_deep_wide_static_threaded/threads:8       6258 ns      49873 ns      14008
```

PS: The implementation could probably be optimized even more.

Differential Revision: [D24300702](https://our.internmc.facebook.com/intern/diff/D24300702)

Commit:
[StaticRuntime] Add a 'speed of light' benchmark.

[ghstack-poisoned]

Commit:
Update on "[StaticRuntime] Add a 'speed of light' benchmark."


This PR adds a hand optimized version of DeepAndWide model with the goal
of estimating overheads of static runtime. While static runtime is
currently much faster than the existing JIT interpreter, it would be
useful to understand how close we are to an absolutely 0-overhead
system.

PS: The implementation could probably be optimized even more.

[ghstack-poisoned]

Commit:
Update on "[StaticRuntime] Add a 'speed of light' benchmark."


This PR adds a hand optimized version of DeepAndWide model with the goal
of estimating overheads of static runtime. While static runtime is
currently much faster than the existing JIT interpreter, it would be
useful to understand how close we are to an absolutely 0-overhead
system. Currently, this "ideal" implementation is 1.5x faster than the
static runtime on batchsize=1.

PS: The implementation could probably be optimized even more.

[ghstack-poisoned]

Commit:
rebase on "[StaticRuntime] Add a 'speed of light' benchmark."


This PR adds a hand optimized version of DeepAndWide model with the goal
of estimating overheads of static runtime. While static runtime is
currently much faster than the existing JIT interpreter, it would be
useful to understand how close we are to an absolutely 0-overhead
system. Currently, this "ideal" implementation is 1.5x faster than the
static runtime on batchsize=1.

PS: The implementation could probably be optimized even more.

[ghstack-poisoned]

Commit:
Update on "[StaticRuntime] Add a 'speed of light' benchmark."


This PR adds a hand optimized version of DeepAndWide model with the goal
of estimating overheads of static runtime. While static runtime is
currently much faster than the existing JIT interpreter, it would be
useful to understand how close we are to an absolutely 0-overhead
system. Currently, this "ideal" implementation is 1.5x faster than the
static runtime on batchsize=1.

PS: The implementation could probably be optimized even more.

Differential Revision: [D24300702](https://our.internmc.facebook.com/intern/diff/D24300702)

[ghstack-poisoned]

Commit:
Update on "[StaticRuntime] Add a 'speed of light' benchmark."


This PR adds a hand optimized version of DeepAndWide model with the goal
of estimating overheads of static runtime. While static runtime is
currently much faster than the existing JIT interpreter, it would be
useful to understand how close we are to an absolutely 0-overhead
system. Currently, this "ideal" implementation is 1.5x faster than the
static runtime on batchsize=1.

PS: The implementation could probably be optimized even more.

Differential Revision: [D24300702](https://our.internmc.facebook.com/intern/diff/D24300702)

[ghstack-poisoned]

dzhulgakov commented on benchmarks/static_runtime/deep_wide_pt.h:
```diff
@@ -31,6 +32,77 @@ struct DeepAndWide : torch::nn::Module {
+      at::native::bmm_out_cpu(
+          prealloc_tensors[4], ad_emb_packed, prealloc_tensors[3]);
+      prealloc_tensors[5] = at::native::flatten(prealloc_tensors[4], 1);
+      at::native::_cat_out_cpu(
```
you can replace `cat` with carefully constructed tensor views on the output that are passed to the `_out` ops above

dzhulgakov commented on benchmarks/static_runtime/deep_wide_pt.h:
```diff
@@ -31,6 +32,77 @@ struct DeepAndWide : torch::nn::Module {
+      prealloc_tensors[3] = at::native::transpose(user_emb, 1, 2);
+      at::native::bmm_out_cpu(
+          prealloc_tensors[4], ad_emb_packed, prealloc_tensors[3]);
+      prealloc_tensors[5] = at::native::flatten(prealloc_tensors[4], 1);
```
`flatten` can be replaced with a view that you pre-allocated

dzhulgakov commented on benchmarks/static_runtime/deep_wide_pt.h:
```diff
@@ -31,6 +32,77 @@ struct DeepAndWide : torch::nn::Module {
+      at::native::clamp_out(
+          prealloc_tensors[2], prealloc_tensors[1], -10.0, 10.0);
+
+      prealloc_tensors[3] = at::native::transpose(user_emb, 1, 2);
```
I'd just prepare pre-transposed view of the original tensor or better yet call MKL directly instead of bmm. Otherwise you create new TensorImpl here

dzhulgakov replied:
you can apply the same trick here as the one I just commented for view/flatten. At the very least do `set_` so that it reuses TensorImpl instance

dzhulgakov commented on benchmarks/static_runtime/deep_wide_pt.h:
```diff
@@ -31,6 +32,77 @@ struct DeepAndWide : torch::nn::Module {
+
+      return pred;
+    } else {
+      at::native::add_out(prealloc_tensors[0], wide, mu_);
+      at::native::mul_out(prealloc_tensors[1], prealloc_tensors[0], sigma_);
```
this could be fused together - I think using Eigen would do the trick. It'd also remove the overhead of TensorIterator that you still invoke today

dzhulgakov reviewed (commented):
Awesome! Can you post complete perf results? And how does it compare to C2 executor?

Commit:
Update on "[StaticRuntime] Add a 'speed of light' benchmark."


This PR adds a hand optimized version of DeepAndWide model with the goal
of estimating overheads of static runtime. While static runtime is
currently much faster than the existing JIT interpreter, it would be
useful to understand how close we are to an absolutely 0-overhead
system. Currently, this "ideal" implementation is 1.5x faster than the
static runtime on batchsize=1.

Full benchmark results:
```
Running build/bin/static_runtime_bench
Run on (24 X 2394.71 MHz CPU s)
CPU Caches:
  L1 Data 32K (x24)
  L1 Instruction 32K (x24)
  L2 Unified 4096K (x24)
  L3 Unified 16384K (x24)
------------------------------------------------------------------------------
Benchmark                                       Time           CPU Iterations
------------------------------------------------------------------------------
BM_deep_wide_base/1                         61450 ns      61432 ns      11038
BM_deep_wide_base/8                         72721 ns      72671 ns       9456
BM_deep_wide_base/20                        75077 ns      75053 ns       9338
BM_deep_wide_fast/1                         17764 ns      17763 ns      39024 << new
BM_deep_wide_fast/8                         26514 ns      26513 ns      26655 << new
BM_deep_wide_fast/20                        28722 ns      28718 ns      23721 << new
BM_deep_wide_jit_graph_executor/1           65673 ns      65657 ns      10925
BM_deep_wide_jit_graph_executor/8           70226 ns      70223 ns       9728
BM_deep_wide_jit_graph_executor/20          73673 ns      73673 ns       9468
BM_deep_wide_jit_profiling_executor/1       63204 ns      63203 ns      10464
BM_deep_wide_jit_profiling_executor/8       71274 ns      71258 ns       9702
BM_deep_wide_jit_profiling_executor/20      75612 ns      75593 ns       9430
BM_deep_wide_static/1                       28436 ns      28434 ns      24496
BM_deep_wide_static/8                       36289 ns      36288 ns      18966
BM_deep_wide_static/20                      39545 ns      39545 ns      17560
BM_deep_wide_static_threaded/threads:8       4150 ns      33184 ns      20472
```

PS: The implementation could probably be optimized even more.

Differential Revision: [D24300702](https://our.internmc.facebook.com/intern/diff/D24300702)

[ghstack-poisoned]

ZolotukhinM commented:
> Can you post complete perf results?

Added to the commit message!

> And how does it compare to C2 executor?

I don't know how to properly measure it in this setting, but I think @bwasti has done investigation of that separately.

As for further optimizations, I probably will take it slow since I need some time to figure out how to actually do them. @dzhulgakov,  is the benchmark valuable in its current form?

bwasti commented on benchmarks/static_runtime/deep_wide_pt.h:
```diff
@@ -31,6 +32,77 @@ struct DeepAndWide : torch::nn::Module {
 
+// Implementation using native functions and pre-allocated tensors.
+// It could be used as a "speed of light" for static runtime.
+struct DeepAndWideFast : torch::nn::Module {
```
might want to check in a test against the non-fast version

ZolotukhinM replied:
wdym?

bwasti commented:
cc @hlu1 about C2 executor perf comparison

dzhulgakov commented on benchmarks/static_runtime/deep_wide_pt_bench.cc:
```diff
@@ -99,6 +114,7 @@ static void BM_deep_wide_static_threaded(benchmark::State& state) {
 }
 
 BENCHMARK(BM_deep_wide_base)->RangeMultiplier(8)->Ranges({{1, 20}});
+BENCHMARK(BM_deep_wide_fast)->RangeMultiplier(8)->Ranges({{1, 20}});
```
while you're on it - you can make this  and the ones below BENCHMARK_RELATIVE then the comparison will look much nicer

ZolotukhinM replied:
I think that's a folly feature, but AFAICT we're not using folly here.

dzhulgakov reviewed (commented):
At least document the possible future improvements as TODOs.

However the preallocated comments are pretty easy to pull off

Commit:
Update on "[StaticRuntime] Add a 'speed of light' benchmark."


This PR adds a hand optimized version of DeepAndWide model with the goal
of estimating overheads of static runtime. While static runtime is
currently much faster than the existing JIT interpreter, it would be
useful to understand how close we are to an absolutely 0-overhead
system. Currently, this "ideal" implementation is 1.5x faster than the
static runtime on batchsize=1.

Full benchmark results:
```
Running build/bin/static_runtime_bench
Run on (24 X 2394.71 MHz CPU s)
CPU Caches:
  L1 Data 32K (x24)
  L1 Instruction 32K (x24)
  L2 Unified 4096K (x24)
  L3 Unified 16384K (x24)
------------------------------------------------------------------------------
Benchmark                                       Time           CPU Iterations
------------------------------------------------------------------------------
BM_deep_wide_base/1                         61450 ns      61432 ns      11038
BM_deep_wide_base/8                         72721 ns      72671 ns       9456
BM_deep_wide_base/20                        75077 ns      75053 ns       9338
BM_deep_wide_fast/1                         17764 ns      17763 ns      39024 << new
BM_deep_wide_fast/8                         26514 ns      26513 ns      26655 << new
BM_deep_wide_fast/20                        28722 ns      28718 ns      23721 << new
BM_deep_wide_jit_graph_executor/1           65673 ns      65657 ns      10925
BM_deep_wide_jit_graph_executor/8           70226 ns      70223 ns       9728
BM_deep_wide_jit_graph_executor/20          73673 ns      73673 ns       9468
BM_deep_wide_jit_profiling_executor/1       63204 ns      63203 ns      10464
BM_deep_wide_jit_profiling_executor/8       71274 ns      71258 ns       9702
BM_deep_wide_jit_profiling_executor/20      75612 ns      75593 ns       9430
BM_deep_wide_static/1                       28436 ns      28434 ns      24496
BM_deep_wide_static/8                       36289 ns      36288 ns      18966
BM_deep_wide_static/20                      39545 ns      39545 ns      17560
BM_deep_wide_static_threaded/threads:8       4150 ns      33184 ns      20472
```

PS: The implementation could probably be optimized even more.

Differential Revision: [D24300702](https://our.internmc.facebook.com/intern/diff/D24300702)

[ghstack-poisoned]

Commit:
Update on "[StaticRuntime] Add a 'speed of light' benchmark."


This PR adds a hand optimized version of DeepAndWide model with the goal
of estimating overheads of static runtime. While static runtime is
currently much faster than the existing JIT interpreter, it would be
useful to understand how close we are to an absolutely 0-overhead
system. Currently, this "ideal" implementation is 1.5x faster than the
static runtime on batchsize=1.

Full benchmark results:
```
Running build/bin/static_runtime_bench
Run on (24 X 2394.71 MHz CPU s)
CPU Caches:
  L1 Data 32K (x24)
  L1 Instruction 32K (x24)
  L2 Unified 4096K (x24)
  L3 Unified 16384K (x24)
------------------------------------------------------------------------------
Benchmark                                       Time           CPU Iterations
------------------------------------------------------------------------------
BM_deep_wide_base/1                         60625 ns      60612 ns      10582
BM_deep_wide_base/8                         81130 ns      81123 ns       8295
BM_deep_wide_base/20                        90470 ns      90470 ns       7246
BM_deep_wide_fast/1                         17926 ns      17923 ns      40064 << new
BM_deep_wide_fast/8                         30111 ns      30110 ns      21388 << new
BM_deep_wide_fast/20                        33236 ns      33228 ns      20716 << new
BM_deep_wide_jit_graph_executor/1           61334 ns      61332 ns       9931
BM_deep_wide_jit_graph_executor/8           79159 ns      79157 ns       9261
BM_deep_wide_jit_graph_executor/20          73963 ns      73957 ns       9225
BM_deep_wide_jit_profiling_executor/1       62241 ns      62240 ns      10843
BM_deep_wide_jit_profiling_executor/8       76708 ns      75937 ns       9283
BM_deep_wide_jit_profiling_executor/20      84358 ns      84055 ns       9247
BM_deep_wide_static/1                       29592 ns      29584 ns      25079
BM_deep_wide_static/8                       42031 ns      42016 ns      16943
BM_deep_wide_static/20                      44553 ns      44383 ns      16242
BM_deep_wide_static_threaded/threads:8       5200 ns      41564 ns      17672
```

PS: The implementation could probably be optimized even more.

Differential Revision: [D24300702](https://our.internmc.facebook.com/intern/diff/D24300702)

[ghstack-poisoned]

ZolotukhinM commented:
> At least document the possible future improvements as TODOs.
> 
> However the preallocated comments are pretty easy to pull off

Added the comments + replaced `flatten` with `view`! However, It didn't noticeably affect the results.

dzhulgakov commented on benchmarks/static_runtime/deep_wide_pt.h:
```diff
@@ -31,6 +32,88 @@ struct DeepAndWide : torch::nn::Module {
+          prealloc_tensors[4], ad_emb_packed, prealloc_tensors[3]);
+
+      prealloc_tensors[5] =
+          prealloc_tensors[4].view({prealloc_tensors[4].size(0), 1});
```
yes, because you still create new TensorImpl every time (it's the same as `flatten`)

What I meant is to re-initialize the tensor using `set_` which allows you to reuse the storage and reset the stride/sizes in-place.

Even better - you can do something like this (it seems correct to me and faster in the default case):

```
if (prealloc_tensors[5].data_ptr() != prealloc_tensors[4].data_ptr()) {
  // in unlikely case that the input tensor changed we need to reinitialize the view
  prealloc_tensors[5] = prealloc_tensors[4].view({prealloc_tensors[4].size(0), 1})
}
```

Commit:
Update on "[StaticRuntime] Add a 'speed of light' benchmark."


This PR adds a hand optimized version of DeepAndWide model with the goal
of estimating overheads of static runtime. While static runtime is
currently much faster than the existing JIT interpreter, it would be
useful to understand how close we are to an absolutely 0-overhead
system. Currently, this "ideal" implementation is 1.5x faster than the
static runtime on batchsize=1.

Full benchmark results:
```
Running build/bin/static_runtime_bench
Run on (24 X 2394.71 MHz CPU s)
CPU Caches:
  L1 Data 32K (x24)
  L1 Instruction 32K (x24)
  L2 Unified 4096K (x24)
  L3 Unified 16384K (x24)
------------------------------------------------------------------------------
Benchmark                                       Time           CPU Iterations
------------------------------------------------------------------------------
BM_deep_wide_base/1                         60625 ns      60612 ns      10582
BM_deep_wide_base/8                         81130 ns      81123 ns       8295
BM_deep_wide_base/20                        90470 ns      90470 ns       7246
BM_deep_wide_fast/1                         17926 ns      17923 ns      40064 << new
BM_deep_wide_fast/8                         30111 ns      30110 ns      21388 << new
BM_deep_wide_fast/20                        33236 ns      33228 ns      20716 << new
BM_deep_wide_jit_graph_executor/1           61334 ns      61332 ns       9931
BM_deep_wide_jit_graph_executor/8           79159 ns      79157 ns       9261
BM_deep_wide_jit_graph_executor/20          73963 ns      73957 ns       9225
BM_deep_wide_jit_profiling_executor/1       62241 ns      62240 ns      10843
BM_deep_wide_jit_profiling_executor/8       76708 ns      75937 ns       9283
BM_deep_wide_jit_profiling_executor/20      84358 ns      84055 ns       9247
BM_deep_wide_static/1                       29592 ns      29584 ns      25079
BM_deep_wide_static/8                       42031 ns      42016 ns      16943
BM_deep_wide_static/20                      44553 ns      44383 ns      16242
BM_deep_wide_static_threaded/threads:8       5200 ns      41564 ns      17672
```

PS: The implementation could probably be optimized even more.

Differential Revision: [D24300702](https://our.internmc.facebook.com/intern/diff/D24300702)

[ghstack-poisoned]

Commit:
Update on "[StaticRuntime] Add a 'speed of light' benchmark."


This PR adds a hand optimized version of DeepAndWide model with the goal
of estimating overheads of static runtime. While static runtime is
currently much faster than the existing JIT interpreter, it would be
useful to understand how close we are to an absolutely 0-overhead
system. Currently, this "ideal" implementation is 1.5x faster than the
static runtime on batchsize=1.

Full benchmark results:
```
Running build/bin/static_runtime_bench
Run on (24 X 2394.71 MHz CPU s)
CPU Caches:
  L1 Data 32K (x24)
  L1 Instruction 32K (x24)
  L2 Unified 4096K (x24)
  L3 Unified 16384K (x24)
------------------------------------------------------------------------------
Benchmark                                       Time           CPU Iterations
------------------------------------------------------------------------------
BM_deep_wide_base/1                         62389 ns      62266 ns      11032
BM_deep_wide_base/8                         76510 ns      76502 ns       7066
BM_deep_wide_base/20                        80194 ns      80192 ns       9050
BM_deep_wide_fast/1                         13995 ns      13992 ns      49349 << new
BM_deep_wide_fast/8                         23757 ns      23756 ns      30340 << new
BM_deep_wide_fast/20                        25284 ns      25273 ns      26551 << new
BM_deep_wide_jit_graph_executor/1           60308 ns      60306 ns      11151
BM_deep_wide_jit_graph_executor/8           69640 ns      69404 ns       9801
BM_deep_wide_jit_graph_executor/20          72013 ns      71992 ns       9439
BM_deep_wide_jit_profiling_executor/1       60420 ns      60420 ns      11097
BM_deep_wide_jit_profiling_executor/8       69242 ns      69222 ns      10038
BM_deep_wide_jit_profiling_executor/20      73953 ns      73952 ns       9189
BM_deep_wide_static/1                       28128 ns      28128 ns      25212
BM_deep_wide_static/8                       36621 ns      36621 ns      19387
BM_deep_wide_static/20                      39560 ns      39551 ns      17866
BM_deep_wide_static_threaded/threads:8       4654 ns      37137 ns      18240
```

PS: The implementation could probably be optimized even more.

Differential Revision: [D24300702](https://our.internmc.facebook.com/intern/diff/D24300702)

[ghstack-poisoned]

Commit:
Update on "[StaticRuntime] Add a 'speed of light' benchmark."


This PR adds a hand optimized version of DeepAndWide model with the goal
of estimating overheads of static runtime. While static runtime is
currently much faster than the existing JIT interpreter, it would be
useful to understand how close we are to an absolutely 0-overhead
system. Currently, this "ideal" implementation is 1.5x faster than the
static runtime on batchsize=1.

Full benchmark results:
```
Running build/bin/static_runtime_bench
Run on (24 X 2394.71 MHz CPU s)
CPU Caches:
  L1 Data 32K (x24)
  L1 Instruction 32K (x24)
  L2 Unified 4096K (x24)
  L3 Unified 16384K (x24)
------------------------------------------------------------------------------
Benchmark                                       Time           CPU Iterations
------------------------------------------------------------------------------
BM_deep_wide_base/1                         62389 ns      62266 ns      11032
BM_deep_wide_base/8                         76510 ns      76502 ns       7066
BM_deep_wide_base/20                        80194 ns      80192 ns       9050
BM_deep_wide_fast/1                         13995 ns      13992 ns      49349 << new
BM_deep_wide_fast/8                         23757 ns      23756 ns      30340 << new
BM_deep_wide_fast/20                        25284 ns      25273 ns      26551 << new
BM_deep_wide_jit_graph_executor/1           60308 ns      60306 ns      11151
BM_deep_wide_jit_graph_executor/8           69640 ns      69404 ns       9801
BM_deep_wide_jit_graph_executor/20          72013 ns      71992 ns       9439
BM_deep_wide_jit_profiling_executor/1       60420 ns      60420 ns      11097
BM_deep_wide_jit_profiling_executor/8       69242 ns      69222 ns      10038
BM_deep_wide_jit_profiling_executor/20      73953 ns      73952 ns       9189
BM_deep_wide_static/1                       28128 ns      28128 ns      25212
BM_deep_wide_static/8                       36621 ns      36621 ns      19387
BM_deep_wide_static/20                      39560 ns      39551 ns      17866
BM_deep_wide_static_threaded/threads:8       4654 ns      37137 ns      18240
```

PS: The implementation could probably be optimized even more.

Differential Revision: [D24300702](https://our.internmc.facebook.com/intern/diff/D24300702)

[ghstack-poisoned]

Commit:
Update on "[StaticRuntime] Add a 'speed of light' benchmark."


This PR adds a hand optimized version of DeepAndWide model with the goal
of estimating overheads of static runtime. While static runtime is
currently much faster than the existing JIT interpreter, it would be
useful to understand how close we are to an absolutely 0-overhead
system. Currently, this "ideal" implementation is 2x faster than the
static runtime on batchsize=1.

Full benchmark results:
```
Running build/bin/static_runtime_bench
Run on (24 X 2394.71 MHz CPU s)
CPU Caches:
  L1 Data 32K (x24)
  L1 Instruction 32K (x24)
  L2 Unified 4096K (x24)
  L3 Unified 16384K (x24)
------------------------------------------------------------------------------
Benchmark                                       Time           CPU Iterations
------------------------------------------------------------------------------
BM_deep_wide_base/1                         62389 ns      62266 ns      11032
BM_deep_wide_base/8                         76510 ns      76502 ns       7066
BM_deep_wide_base/20                        80194 ns      80192 ns       9050
BM_deep_wide_fast/1                         13995 ns      13992 ns      49349 << new
BM_deep_wide_fast/8                         23757 ns      23756 ns      30340 << new
BM_deep_wide_fast/20                        25284 ns      25273 ns      26551 << new
BM_deep_wide_jit_graph_executor/1           60308 ns      60306 ns      11151
BM_deep_wide_jit_graph_executor/8           69640 ns      69404 ns       9801
BM_deep_wide_jit_graph_executor/20          72013 ns      71992 ns       9439
BM_deep_wide_jit_profiling_executor/1       60420 ns      60420 ns      11097
BM_deep_wide_jit_profiling_executor/8       69242 ns      69222 ns      10038
BM_deep_wide_jit_profiling_executor/20      73953 ns      73952 ns       9189
BM_deep_wide_static/1                       28128 ns      28128 ns      25212
BM_deep_wide_static/8                       36621 ns      36621 ns      19387
BM_deep_wide_static/20                      39560 ns      39551 ns      17866
BM_deep_wide_static_threaded/threads:8       4654 ns      37137 ns      18240
```

PS: The implementation could probably be optimized even more.

Differential Revision: [D24300702](https://our.internmc.facebook.com/intern/diff/D24300702)

[ghstack-poisoned]

ZolotukhinM commented:
@dzhulgakov, I've done the two changes you mentioned, the first one (around `view`) clearly boosted the performance, but I'm not sure that I did the second one right (around `transpose`), since it didn't have a noticeable effect. Please take a look! All in all, this is now 2x faster than even static runtime.

dzhulgakov commented on benchmarks/static_runtime/deep_wide_pt.h:
```diff
@@ -31,6 +32,99 @@ struct DeepAndWide : torch::nn::Module {
+      // prealloc_tensors[3] = at::native::transpose(user_emb, 1, 2);
+      auto sizes = user_emb.sizes();
+      auto strides = user_emb.strides();
+      prealloc_tensors[3].set_(
```
yeah, it seems correct - maybe because we still call `set_` and go through a bunch of field assignments? (even though TensorImpl is the same). You can probably wrap it into the same `if (prealloc_tensors[3].data_ptr() != user_emb.data_ptr())` as below and see whether it helps

dzhulgakov reviewed (approved):
Let's get it in then! We could still optimize `cat` and unwrap other methods to not do shape checking, but we should start somewhere :)

Commit:
Update on "[StaticRuntime] Add a 'speed of light' benchmark."


This PR adds a hand optimized version of DeepAndWide model with the goal
of estimating overheads of static runtime. While static runtime is
currently much faster than the existing JIT interpreter, it would be
useful to understand how close we are to an absolutely 0-overhead
system. Currently, this "ideal" implementation is 2x faster than the
static runtime on batchsize=1.

Full benchmark results:
```
Running build/bin/static_runtime_bench
Run on (24 X 2394.71 MHz CPU s)
CPU Caches:
  L1 Data 32K (x24)
  L1 Instruction 32K (x24)
  L2 Unified 4096K (x24)
  L3 Unified 16384K (x24)
------------------------------------------------------------------------------
Benchmark                                       Time           CPU Iterations
------------------------------------------------------------------------------
BM_deep_wide_base/1                         59518 ns      59500 ns      10909
BM_deep_wide_base/8                         74635 ns      74632 ns       9317
BM_deep_wide_base/20                        82186 ns      82147 ns       9119
BM_deep_wide_fast/1                         13851 ns      13851 ns      49825 << new
BM_deep_wide_fast/8                         22497 ns      22497 ns      32089 << new
BM_deep_wide_fast/20                        23868 ns      23841 ns      31184 << new
BM_deep_wide_jit_graph_executor/1           62786 ns      62786 ns      10835
BM_deep_wide_jit_graph_executor/8           76730 ns      76718 ns       7529
BM_deep_wide_jit_graph_executor/20          78886 ns      78883 ns       8769
BM_deep_wide_jit_profiling_executor/1       69504 ns      69490 ns      10309
BM_deep_wide_jit_profiling_executor/8       75718 ns      75715 ns       9199
BM_deep_wide_jit_profiling_executor/20      75364 ns      75364 ns       9010
BM_deep_wide_static/1                       40324 ns      40318 ns      17232
BM_deep_wide_static/8                       50327 ns      50319 ns      13335
BM_deep_wide_static/20                      53075 ns      53071 ns      12855
BM_deep_wide_static_threaded/threads:8       6258 ns      49873 ns      14008
```

PS: The implementation could probably be optimized even more.

Differential Revision: [D24300702](https://our.internmc.facebook.com/intern/diff/D24300702)

[ghstack-poisoned]

Commit:
Update on "[StaticRuntime] Add a 'speed of light' benchmark."


This PR adds a hand optimized version of DeepAndWide model with the goal
of estimating overheads of static runtime. While static runtime is
currently much faster than the existing JIT interpreter, it would be
useful to understand how close we are to an absolutely 0-overhead
system. Currently, this "ideal" implementation is 2x faster than the
static runtime on batchsize=1.

Full benchmark results:
```
Running build/bin/static_runtime_bench
Run on (24 X 2394.71 MHz CPU s)
CPU Caches:
  L1 Data 32K (x24)
  L1 Instruction 32K (x24)
  L2 Unified 4096K (x24)
  L3 Unified 16384K (x24)
------------------------------------------------------------------------------
Benchmark                                       Time           CPU Iterations
------------------------------------------------------------------------------
BM_deep_wide_base/1                         59518 ns      59500 ns      10909
BM_deep_wide_base/8                         74635 ns      74632 ns       9317
BM_deep_wide_base/20                        82186 ns      82147 ns       9119
BM_deep_wide_fast/1                         13851 ns      13851 ns      49825 << new
BM_deep_wide_fast/8                         22497 ns      22497 ns      32089 << new
BM_deep_wide_fast/20                        23868 ns      23841 ns      31184 << new
BM_deep_wide_jit_graph_executor/1           62786 ns      62786 ns      10835
BM_deep_wide_jit_graph_executor/8           76730 ns      76718 ns       7529
BM_deep_wide_jit_graph_executor/20          78886 ns      78883 ns       8769
BM_deep_wide_jit_profiling_executor/1       69504 ns      69490 ns      10309
BM_deep_wide_jit_profiling_executor/8       75718 ns      75715 ns       9199
BM_deep_wide_jit_profiling_executor/20      75364 ns      75364 ns       9010
BM_deep_wide_static/1                       40324 ns      40318 ns      17232
BM_deep_wide_static/8                       50327 ns      50319 ns      13335
BM_deep_wide_static/20                      53075 ns      53071 ns      12855
BM_deep_wide_static_threaded/threads:8       6258 ns      49873 ns      14008
```

PS: The implementation could probably be optimized even more.

Differential Revision: [D24300702](https://our.internmc.facebook.com/intern/diff/D24300702)

[ghstack-poisoned]
